src/atlas/redistribution/detail/RedistributeGeneric.cc

```diff
@@ -0,0 +1,498 @@
+    return *functionSpaceImpl;
+}
+
+// Get ghost field from functionspace.
+template <typename FunctionSpaceType, HasGhost<FunctionSpaceType> = nullptr>
+Field getGhostField( const FunctionSpaceType* functionSpaceImpl ) {
+    return functionSpaceImpl->ghost();
+}
+
+// Create ghost field.
+template <typename FunctionSpaceType, NoGhost<FunctionSpaceType> = nullptr>
+Field getGhostField( const FunctionSpaceType* functionSpaceImpl ) {
+
+    // Get mesh elements.
+    const auto& elems = getElems( functionSpaceImpl );
+
+    // Make ghost field.
+    auto ghost = Field( "ghost" , array::make_datatype<int>(),
+                        array::make_shape( functionSpaceImpl->size() ) );
+
+    // Make views.
+    auto ghostView = array::make_view<int, 1>( ghost );
+    auto remoteIndexView = array::make_indexview<idx_t, 1>( elems.remote_index() );
+    auto partitionView = array::make_view<int, 1>( elems.partition() );
+
+    // Set ghost field.
+    const auto thisPart = static_cast<int>( mpi::comm().rank() );
+    for ( idx_t i = 0; i < ghost.shape( 0 ); ++i ) {
+        ghostView( i ) = partitionView( i ) != thisPart ||
+                         remoteIndexView ( i ) != i;
+    }
+    return ghost;
+}
+
+// Get UID field from functionspace global indices.
+template <typename FunctionSpaceType, HasGidx<FunctionSpaceType> = nullptr>
+Field getUidField( const FunctionSpaceType* functionSpaceImpl ) {
+    return getElems( functionSpaceImpl ).global_index();
+}
+
+// Create UID field.
+template <typename FunctionSpaceType, NoGidx<FunctionSpaceType> = nullptr>
+Field getUidField( const FunctionSpaceType* functionSpaceImpl ) {
+
+    // Make UID field.
+    Field uid = Field( " Uniquie ID ", array::make_datatype<uidx_t>(),
+                       array::make_shape( functionSpaceImpl->size() ) );
+
+    // Make views.
+    auto uidView = array::make_view<uidx_t, 1>( uid );
+    const auto lonLatView = array::make_view<double, 2>( functionSpaceImpl->lonlat() );
+
+    // Set UIDs
+    for ( idx_t i = 0; i < uid.shape(0); ++i ) {
+        uidView( i ) = util::unique_lonlat( lonLatView (i, LON), lonLatView(i, LAT ) );
+    }
+    return uid;
+}
+
+// Get vector of (local index, UID) pairs.
+template <typename FunctionSpaceType>
+std::vector<IdxUid> getUidVec(const FunctionSpaceType* functionSpaceImpl) {
+
+    // Get ghost and unique ID fields from functionspace.
+    const auto ghost = getGhostField( functionSpaceImpl );
+    const auto uid = getUidField( functionSpaceImpl );
+
+    // Make views to fields.
+    const auto ghostView = array::make_view<int, 1>( ghost );
+    const auto uidView = array::make_view<uidx_t, 1>( uid );
+
+    auto uidVec = std::vector<IdxUid>{};
+
+    // Get UIDs for non ghost elems.
+    for ( idx_t i = 0; i < functionSpaceImpl->size(); ++i ) {
+        if ( ghostView( i ) ) {
+            continue;
+        }
+       uidVec.push_back( IdxUid( i, uidView( i ) ) );
+    }
```

Comment:
And this could be instead:
```
if( not ghostView(i) ) {
    uidVec.emplace_back( i, uidView(i) );
}
```
`emplace_back` has variadic arguments that forwards to the constructor and does an in-place construction, rather than the original copy (or perhaps move) construction